src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,246 @@
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"io"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/digest"
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/storage/index/convert"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/dbnode/x/xio"
+	"github.com/m3db/m3/src/x/checked"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+type Merger struct {
+	reader        DataFileSetReader
+	srPool        xio.SegmentReaderPool
+	multiIterPool encoding.MultiReaderIteratorPool
+	identPool     ident.Pool
+	encoderPool   encoding.EncoderPool
+}
+
+func NewMerger(
+	reader DataFileSetReader,
+	srPool xio.SegmentReaderPool,
+	multiIterPool encoding.MultiReaderIteratorPool,
+	identPool ident.Pool,
+	encoderPool encoding.EncoderPool,
+) *Merger {
+	return &Merger{
+		reader:        reader,
+		srPool:        srPool,
+		multiIterPool: multiIterPool,
+		identPool:     identPool,
+		encoderPool:   encoderPool,
+	}
+}
+
+func (m *Merger) Merge(
+	mergeWith MergeWith,
+	ns namespace.Metadata,
+	shard uint32,
+	blockStart xtime.UnixNano,
+	blockSize time.Duration,
+	flushPreparer persist.FlushPreparer,
+	nsCtx namespace.Context,
+) error {
+	var multiErr xerrors.MultiError
+
+	reader := m.reader
+	srPool := m.srPool
+	multiIterPool := m.multiIterPool
+	identPool := m.identPool
+	encoderPool := m.encoderPool
+
+	startTime := blockStart.ToTime()
+	openOpts := DataReaderOpenOptions{
+		Identifier: FileSetFileIdentifier{
+			Namespace:  ns.ID(),
+			Shard:      shard,
+			BlockStart: startTime,
+		},
+	}
+	if err := reader.Open(openOpts); err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+	defer reader.Close()
+
+	prepareOpts := persist.DataPrepareOptions{
+		NamespaceMetadata: ns,
+		Shard:             shard,
+		BlockStart:        startTime,
+		DeleteIfExists:    false,
+	}
+	prepared, err := flushPreparer.PrepareData(prepareOpts)
+	if err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+
+	// Writing data for a block is done in two stages. The first stage is
+	// to loop through series on disk and merge it with what's in the merge
+	// target. The second stage is to persist the rest of the series in the
+	// merge target that was not persisted in the first stage.
+
+	// First stage: loop through series on disk.
+	for id, tagsIter, data, _, err := reader.Read(); err != io.EOF; id, tagsIter, data, _, err = reader.Read() {
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+
+		// There will be at most two BlockReader slices: one for disk data
+		// and one for in memory data.
+		brs := make([][]xio.BlockReader, 0, 2)
+
+		// Create BlockReader slice out of disk data.
+		br := blockReaderFromData(data, srPool, startTime, blockSize)
```

Comment:
You get a segment reader from the pool each time this function is called, but do you ever return it? If not you should probably do that.

The more efficient alternative would be to just reuse the same two segment readers over and over again (they have a `Reset()` method)

Reply:
👍 will do the more efficient alternative.